src/main/java/org/truffleruby/core/array/ArrayNodes.java

```diff
@@ -1891,32 +1895,61 @@ public DynamicObject sortVeryShort(VirtualFrame frame, DynamicObject array, NotP
             return createArray(store.getArray(), size);
         }
 
-        @Specialization(guards = { "!isEmptyArray(array)", "!isSmall(array)" })
-        public Object sortLargeArray(VirtualFrame frame, DynamicObject array, NotProvided block,
-                @Cached("new()") SnippetNode snippetNode) {
-            return snippetNode.execute(frame,
-                    "sorted = dup; Truffle.privately { sorted.mergesort! }; sorted");
+        @Specialization(guards = { "!isEmptyArray(array)", "!isSmall(array)", "isIntArray(array)" },
+                        assumptions = "fixnumCmpAssumption")
+        public Object sortIntArray(DynamicObject array, NotProvided block) {
+            final int size = getSize(array);
+            int[] copy = ((int[]) getStore(array)).clone();
+            doSort(copy, size);
+            return createArray(copy, size);
         }
 
-        @Specialization(guards = { "!isEmptyArray(array)", "isObjectArray(array)" })
-        public Object sortObjectWithBlock(DynamicObject array, DynamicObject block) {
+        @Specialization(guards = { "!isEmptyArray(array)", "!isSmall(array)", "isLongArray(array)" },
+                        assumptions = "fixnumCmpAssumption")
+        public Object sortLongArray(DynamicObject array, NotProvided block) {
             final int size = getSize(array);
-            Object[] copy = ((Object[]) getStore(array)).clone();
-            doSort(copy, size, block);
+            long[] copy = ((long[]) getStore(array)).clone();
+            doSort(copy, size);
             return createArray(copy, size);
         }
 
+        @Specialization(guards = { "!isEmptyArray(array)", "!isSmall(array)", "isDoubleArray(array)" },
+                        assumptions = "floatCmpAssumption")
+        public Object sortDoubleArray(DynamicObject array, NotProvided block) {
+            final int size = getSize(array);
+            double[] copy = ((double[]) getStore(array)).clone();
+            doSort(copy, size);
+            return createArray(copy, size);
+        }
+
+        @Specialization(guards = "!isEmptyArray(array)")
+        public Object sortGenericWithBlock(VirtualFrame frame, DynamicObject array, DynamicObject block,
+                                           @Cached("new()") SnippetNode snippetNode) {
+            return snippetNode.execute(frame,
+                    "sorted = dup; Truffle.privately { sorted.mergesort_block!(block) }; ret = []; Truffle::Array.steal_storage(ret, sorted)",
+                   "block", block);
+        }
+
+        @Specialization(guards = "!isEmptyArray(array)")
+        public Object sortGenericWithoutBlock(VirtualFrame frame, DynamicObject array, NotProvided block,
```

Comment:
Can you swap with the Specialization above? So the specializations without blocks are together, and this is clearly the fallback for specialized int[], etc (when the assumption doesn't hold).

Reply:
Sure. And that does bring up another question about fallbacks. I couldn't find a way to make `@Fallback` work due to the first specialization using a cached value. But as written, the specialization order is really important. If you move these specializations about the primitive array ones I just added, you won't get the fast operations. I suppose I could use `replaces`, but I also want to handle the `Object[]` case with these "generic" specializations. Is there a more robust way to achieve what I want?

Reply:
`@Fallback` wouldn't work well here since we at least want to differentiate with and without block I think here.
You could use `replaces`, it's a decision of whether it's worth keeping the previous specializations if we none of them match (probably yes here).

In this case, it might good to differentiate the `Object[]` case and the fallback for primitive arrays (when the Assumption doesn't hold).
Then the `Object[]` case should have guards so it doesn't get picked up for a primitive array, and so the `Arrays.sort` specializations are always chosen for primitives arrays, until <=> is redefined where they'd go to their fallback.

Reply:
The issue there is the `Object[]` and the primitive fallbacks would be the same. And on assumption invalidation, I wouldn't want to retain the original specializations because they'd rely on an assumption about `<=>` that no longer holds.

But I guess I can separate them and then just delegate to a private method.

Reply:
> The issue there is the `Object[]` and the primitive fallbacks would be the same.

It's not an issue, just extract a method :)

> And on assumption invalidation, I wouldn't want to retain the original specializations because they'd rely on an assumption about `<=>` that no longer holds.

The DSL can handle that itself, it will remove specializations with invalid assumptions, and not try to instantiate them if the Assumption is invalid.
Fixnum#<=>  might be redefined but not Float#<=>, so it's better to not remove them eagerly with `replaces` here.

Reply:
> It's not an issue, just extract a method :)

Fair enough. I didn't want to have mostly duplicated specializations. But, so be it.

> The DSL can handle that itself, it will remove specializations with invalid assumptions, and not try to instantiate them if the Assumption is invalid.

I mentioned it because I was confused by you saying "whether it's worth keeping the previous specializations." But I see now you just meant across int/long/float.